sql: add get_bit() and set_bit() builtin functions for bits

abhishek20123g opened the pull request:
refs #45851

This commit adds builtin function get_bit() and set_bit() for varbits,
along with thier respective testcases.

Characteristics of the above builtin functions are:
    * get_bit() allows us to extracts a bit at given index in
      the bit array.
    * set_bit() enable us updates a bit at given index in the
      bit array.

Release justification: Release justification: low-risk addition of builtin functions.

Release note (sql change): This PR is introduced to add get_bits()
and set_bit() builtin functions for bits.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45957)
<!-- Reviewable:end -->

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -321,6 +321,53 @@ var builtins = map[string]builtinDefinition{
+	// https://www.postgresql.org/docs/9.0/functions-binarystring.html#FUNCTIONS-BINARYSTRING-OTHER
+	"get_bit": makeBuiltin(tree.FunctionProperties{Category: categoryString},
+		tree.Overload{
+			Types:      tree.ArgTypes{{"val", types.VarBit}, {"position", types.Int}},
```
I'd rename this argument to bits, or bitString

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -321,6 +321,53 @@ var builtins = map[string]builtinDefinition{
+					return nil, pgerror.Newf(
+						pgcode.ArraySubscript, "bit index %d out of valid range (0..%d)", position, val.BitLen()-1)
+				}
+				return tree.NewDInt(tree.DInt(val.BitArray.String()[position] - '0')), nil
```
[nit], but I find would find it clearer to case on the value in the bit array rather than doing a string cast into character math

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -321,6 +321,53 @@ var builtins = map[string]builtinDefinition{
+				}
+				return tree.NewDInt(tree.DInt(val.BitArray.String()[position] - '0')), nil
+			},
+			Info: "extract a particular bit from a given array of bits",
```
[nit]: extract a bit at a given position...

yuzefovich commented:
Putting @rohany and @otan as reviewers since you were suggested by GitHub :)

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -321,6 +321,53 @@ var builtins = map[string]builtinDefinition{
+				}
+				// Converted BitArray to rune format, and update the given index.
+				encodedString := []rune(val.BitArray.String())
+				encodedString[position] = rune('0' + toSet)
```
For this operation and the above operation, I think it would be worth it to implement a get and set operation on a `bitarray.BitArray` rather than doing these various casts to strings and back.

abhishek20123g replied:
Sure, 
I was thinking of doing the same because doing it by casting quite was unnatural.
along with this, I had broken change into 2 commits one for adding methods in ```bitarray.BitArray``` and another for adding ```builtin functions```

abhishek20123g replied:
hii @rohany, 
my changes are not able to pass TeamCity build failed, I had tried to debug but can not able to narrow down the cause, log report generated says 
``` Process exited with code 2 (Step: Verify source tarball (Command Line))
  Step Verify source tarball (Command Line) failed
Step 4/4: Assert Workspace Clean (Command Line)````
and 
```  Process exited with code 1 (Step: Assert Workspace Clean (Command Line))
  Step Assert Workspace Clean (Command Line) failed```
can you point me where to log for debugging 

thanks

rohany replied:
I see this error in the output:
```
[Test archive] fatal: Not a git repository (or any of the parent directories): .git
[14:38:05]	[Test archive] github.com/cockroachdb/cockroach/pkg/util/bitarray
[14:38:05]	[Test archive] # github.com/cockroachdb/cockroach/pkg/util/bitarray
[14:38:05]	[Test archive] pkg/util/bitarray/bitarray.go:552:35: invalid operation: uint64(1) << (numBitsPerWord - 1 - index % numBitsPerWord) (shift count type int, must be unsigned integer)
[14:38:05]	[Test archive] pkg/util/bitarray/bitarray.go:570:38: invalid operation: uint64(1) << (numBitsPerWord - 1 - index % numBitsPerWord) (shift count type int, must be unsigned integer)
[14:38:05]	[Test archive] pkg/util/bitarray/bitarray.go:572:49: invalid operation: uint64(toSet) << (numBitsPerWord - 1 - index % numBitsPerWord) (shift count type int, must be unsigned integer)
[14:38:05]	[Test archive] make[1]: *** No rule to make target 'bin/execgen', needed by 'pkg/col/coldata/vec.eg.go'.  Stop.
[14:38:05]	[Test archive] make[1]: *** Waiting for unfinished jobs....
```

otan commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -321,6 +321,53 @@ var builtins = map[string]builtinDefinition{
+			Types: tree.ArgTypes{
+				{"val", types.VarBit},
+				{"position", types.Int},
+				{"toSet", types.Int},
```
also minor nit: `to_set` (we generally use snake case)

abhishek20123g replied:
Sure,

abhishek20123g committed:
util: add GetBitAtIndex and SetBitAtIndex method in bitarray.BitArray

This commit adds GetBitAtIndex and SetBitAtIndex along with thier
respective testcases in bitarray.BitArray

Characteristics of the above methods are:
    * GetBitAtIndex() extract bit at given index in the BitArray.
    * SetBitAtIndex() returns the BitArray with an updated bit
      at a given index.

Release note: None

abhishek20123g commented on pkg/util/bitarray/bitarray.go:
```diff
@@ -538,3 +538,37 @@ func Next(d BitArray) BitArray {
+}
+
+// SetBitAtIndex returns the BitArray with an updated bit at a given index.
+func (d BitArray) SetBitAtIndex(index, toSet int) (BitArray, error) {
```
hi @rohany,
I am had tried SetBitAtIndex method without a deep copy, it caused a problem in UNION operation,

as when I did ```SELECT set_bit(B'1101010', 0, 0) UNION SELECT set_bit(B'1101010', 2, 1)```
for this test case it only generated output ```0111010```
single row only

otan replied:
we should deep copy, otherwise something like 

```
otan=# create table ba_test(a varbit);
CREATE TABLE
otan=# insert into ba_test values ('00');
INSERT 0 1
otan=# select set_bit(a, 1 ,1), get_bit(a, 1), set_bit(a, 1, 0), get_bit(a, 1) from ba_test;
 set_bit | get_bit | set_bit | get_bit
---------+---------+---------+---------
 01      |       0 | 00      |       0
(1 row)
```
would get inconsistent results.

otan reviewed (commented):
are there any plans to get this to work with `::bytea`?

abhishek20123g commented:
> are there any plans to get this to work with ::bytea?
@otan I think I can make it work for ::bytea by adding separete overload

abhishek20123g commented on pkg/sql/logictest/testdata/logic_test/builtin_function:
```diff
@@ -2563,3 +2563,28 @@ query T
+
+subtest get_bit
+
+query I
```
Sure @rohany ,
I will add ```rowsort``` directive too just wondering what does ```rowsort``` signify?

rohany replied:
It allows the test to still verify correctness if the rows are out of order -- there is no ordering gaurantee on the rows from `UNION`

abhishek20123g commented:
hi @otan and @rohany ,  
I need a little advice about which method to adopt:
1. write separate logic for ```bytea``` for get_bit and set_bit inside thier Overloads.
2. Typecast ```bytea``` to ```varbit``` and use the previously written functions.
as the 1st approach is little faster than 2nd as for set_bit typecasting will be 2 times, but the 2nd approach have clean and less redundant code.
please can you tell me which method to adopt?

otan commented:
looks like postgres does not do the typecast but defines it's own for both bit and bytea:

```
otan=# select oid, proname, proargtypes, proargnames from pg_proc where proname = 'set_bit';
 oid  | proname | proargtypes | proargnames
------+---------+-------------+-------------
  724 | set_bit | 17 23 23    |
 3033 | set_bit | 1560 23 23  |
(2 rows)
```

so i think having two is okay.

if you wanted this in two separate PRs, that is okay, just change your PR message to "refs" instead of "resolves", change the title to include "for bits" and implement the "bytea" implementation in another PR

abhishek20123g commented:
hi @otan,
I can have both of them in one PR only If it does cause any problem.
I had almost completed implementation just writing test cases.

abhishek20123g commented:
hi @otan,
Sorry for not being able to add it for  ```bytea```
I did something wrong with my previous implementation which may require sometime correct, therefore I have decided to separate it in two PR. I will create PR for ```bytea``` tmrr.

otan commented:
so is this one ready for review?

abhishek20123g commented:
sorry I have pushed without changes

abhishek20123g commented:
hi @otan and @rohany,
Can you please review, changes,
recently I had revisited commits and some changes:
1. Added testcase for when bit array length is 0 and when bit array if passed as `::varbit`.
 ```
query T rowsort
SELECT set_bit('000000'::varbit, 5, 1) UNION SELECT set_bit('111111'::varbit, 5, 0)
```
```
 query error set_bit\(\): SetBitAtIndex: bit index 0 out of valid range \(0..-1\)
SELECT set_bit(B'', 0, 1)
```
2. Changed 
```
err := fmt.Errorf("GetBitAtIndex: bit index %d out of valid range (0..%d)", index, d.BitLen()-1)
```
to 
```
err := fmt.Errorf("GetBitAtIndex: bit index %d out of valid range (0..%d)", index, int(d.BitLen())-1)
```
as it was giving result.
```
root@localhost:26257/defaultdb> select get_bit(B'', 0);
ERROR: get_bit(): GetBitAtIndex: bit index 0 out of valid range (0..18446744073709551615)
SQLSTATE: 2202E
```
now it is changed to 
```
root@localhost:26257/defaultdb> select get_bit(B'', 0);
ERROR: get_bit(): GetBitAtIndex: bit index 0 out of valid range (0..-1)
SQLSTATE: 2202E
```
which is same in psql.
```
abhishek=> select get_bit(B'', 0);
ERROR:  bit index 0 out of valid range (0..-1)
```
3. changed error message when `to_set` is other than 1 and 0 to.
   to ```new bit must be 0 or 1.``` which is same as psql.

soon I will add PR for bytea :blush:

abhishek20123g committed:
sql: add get_bit() and set_bit() builtin functions for bits.
refs https://github.com/cockroachdb/cockroach/issues/45851

This commit adds builtin function get_bit() and set_bit() for varbits,
along with thier respective testcases.

Characteristics of the above builtin functions are:
    * get_bit() allows us to extracts a bit at given index in
      the bit array.
    * set_bit() enable us updates a bit at given index in the
      bit array.

Release justification: low-risk addition of builtin functions.

Release note (sql change): This PR is introduced to add get_bits()
and set_bit() builtin functions for bits.

otan reviewed (approved):
LGTM, thank you for the attention to detail with postgres impl for those error messages!

otan commented:
bors r+
